Admin console log levels.

gokaysatir opened the pull request:
Now chosen log level is propagated to forkit and kits.
Also, admin console users can filter logs according to their channel names on client side.

Change-Id: Ife15a6148ed87533b81e9d63da252c633e74e559


* Resolves: # <!-- related github issue -->
* Target version: master 

### Summary


### TODO

- [x] ...

### Checklist

- [x] Code is properly formatted
- [x] All commits have Change-Id
- [ ] I have run tests with `make check`
- [x] I have issued `make run` and manually verified that everything looks okay
- [x] Documentation (manuals or wiki) has been updated or is not required

Ashod commented on kit/ForKit.cpp:
```diff
@@ -146,6 +146,12 @@ class ServerWSHandler final : public WebSocketHandler
+        {
+            // Set environment variable so that new children will also set their log levels accordingly.
+            setenv("LOOL_LOGLEVEL", tokens[1].c_str(), true);
+			Log::logger().setLevel(tokens[1]);
```
Please don't mix tabs and spaces. All C and C++ files use spaces (and never tabs) in this project.

Ashod commented on wsd/Admin.cpp:
```diff
@@ -603,42 +603,34 @@ unsigned Admin::getNetStatsInterval()
-
     std::string levelList[9] = {"none", "fatal", "critical", "error", "warning", "notice", "information", "debug", "trace"};
+    unsigned int wsdLogLevel = Log::logger().get("wsd").getLevel();
+	std::string result = "wsd=" + levelList[wsdLogLevel];
```
Tab -> space.

Ashod commented on wsd/Admin.cpp:
```diff
@@ -603,42 +603,34 @@ unsigned Admin::getNetStatsInterval()
     Log::logger().names(nameList);
 
-    for (size_t i = 0; i < nameList.size(); i++)
+    std::string levelList[9] = {"none", "fatal", "critical", "error", "warning", "notice", "information", "debug", "trace"};
```
This is duplicated in two function. Let's just have one please and use it in both functions.

Also, it should be static const and there is no reason to specify the size of the array (it's better to let the compiler figure out how many entries there are).

Ashod commented on wsd/Admin.cpp:
```diff
@@ -603,42 +603,34 @@ unsigned Admin::getNetStatsInterval()
-            break;
-        }
+        LOOLWSD::setLogLevelsOfKits(level); // For current kits.
+		LOOLWSD::sendMessageToForKit("setloglevel " + level); // For forkit and future kits.
```
Many tabs, here and elsewhere.

Ashod commented on wsd/Admin.hpp:
```diff
@@ -173,6 +173,7 @@ class Admin : public SocketPoll
     uint64_t _lastRecvCount;
     size_t _totalSysMemKb;
     size_t _totalAvailMemKb;
+	std::string _forkitLogLevel = "";
```
Is there a reason the constructor can't be trusted to initialize the string?

Ashod commented on wsd/DocumentBroker.hpp:
```diff
@@ -230,6 +230,8 @@ class DocumentBroker : public std::enable_shared_from_this<DocumentBroker>
         alertAllUsers("error: cmd=" + cmd + " kind=" + kind);
     }
 
+	void setLogLevel(const std::string& level);
```
Documentation is helpful to others to understand what this is supposed to do. Also, the name is misleading. If this is supposed to set the *child* process's log level, it should be setChildLogLevel or setKitLogLevel. But regardless of the name, please document all new functions, even if their purpose seems obvious to you.

Ashod commented on wsd/LOOLWSD.cpp:
```diff
@@ -1791,6 +1791,20 @@ void LOOLWSD::autoSave(const std::string& docKey)
     }
 }
 
+void LOOLWSD::setLogLevelsOfKits(const std::string& level) {
```
The C++ code in this project uses braces on new lines. Please keep the code consistent.

Ashod commented on wsd/LOOLWSD.cpp:
```diff
@@ -1791,6 +1791,20 @@ void LOOLWSD::autoSave(const std::string& docKey)
+
+    LOG_INF("Changing kits' log levels: [" << level << ']');
+
+    for (auto& brokerIt : DocBrokers)
```
const auto& ?

Ashod commented on wsd/LOOLWSD.hpp:
```diff
@@ -400,6 +400,8 @@ class LOOLWSD : public Poco::Util::ServerApplication
     /// Autosave a given document (currently only called from Admin).
     static void autoSave(const std::string& docKey);
 
+    static void setLogLevelsOfKits(const std::string& level);
```
Documentation please.

Ashod commented on loleaflet/admin/adminLog.html:
```diff
@@ -9,7 +9,18 @@ <h3 class="title is-3"><script>document.write(l10nstrings.strLog)</script></h3>
         <script>document.write(l10nstrings.strSetLogLevels)</script>
       </button>
-    </div>
+	</div>
```
It seems this file isn't using tabs either. Please review your editor settings, it seems to prefer tabs, which isn't really very friendly (because it depends on the settings of each user's editor). Prefer spaces, unless there is a good reason not to.

Ashod reviewed (commented):
Thanks Gökay. Looks useful.

I left some feedback.

gokaysatir commented:
@Ashod thanks for the comments :) I indeed use VS Code but it sometimes confuses where to put tabs / spaces. So i had switched that feature off. Now i turned it on again. Sorry for too many tabs/spaces confusion. I updated other parts as you pointed out. Thank you.

Ashod commented on kit/ForKit.cpp:
```diff
@@ -150,6 +150,12 @@ class ServerWSHandler final : public WebSocketHandler
+        else if (tokens.size() == 2 && tokens.equals(0, "setloglevel"))
+        {
+            // Set environment variable so that new children will also set their log levels accordingly.
+            setenv("LOOL_LOGLEVEL", tokens[1].c_str(), true);
```
The second param to setenv is an int, not a bool (it's a C function). Best not to mix types (mostly to avoid confusing maintainers).

gokaysatir replied:
Done.

Ashod commented on loleaflet/admin/src/AdminSocketLog.js:
```diff
@@ -100,11 +103,64 @@ var AdminSocketLog = AdminSocketBase.extend({
+		}
+
+		// Does previously selected channel name still exist? If not, we will re-set filter.
+		if (!channelList.includes(currentFilteredChannel))
```
I think all JS code (in this project) use braces for all code blocks. Best to stay consistent, if that's the case. Same for "else" below.

gokaysatir replied:
It is also preferred in Javascript.

Ashod replied:
Yes, so why not add braces here?

Ashod commented on wsd/Admin.cpp:
```diff
@@ -603,42 +603,32 @@ unsigned Admin::getNetStatsInterval()
 
-    for (size_t i = 0; i < nameList.size(); i++)
+    if (std::find(std::begin(levelList), std::end(levelList), level) == std::end(levelList))
+        level = "trace";
```
Should we really default to trace? I expect information (or debug, at most). Trace is only used for troubleshooting and not recommended as a default. See loolwsd.xml for the default config value please.

gokaysatir replied:
It is a fallback rather than default. It should never get hit. I changed it to debug.

Ashod commented on wsd/Admin.cpp:
```diff
@@ -603,42 +603,32 @@ unsigned Admin::getNetStatsInterval()
-        if (nameList[i] == _channelName)
-        {
-            Log::logger().get(nameList[i]).setLevel(_level);
-            break;
-        }
+        LOOLWSD::setLogLevelsOfKits(level); // For current kits.
```
Side note: we refer to the kit instances in the singular because we think in terms of each document being in its own process. (Also, the plural is awkward.)
```
$ git grep -i "kit " | wc -l
393

$ git grep -i "kits " | wc -l
12
```

gokaysatir replied:
Renamed.

Ashod commented on wsd/DocumentBroker.cpp:
```diff
@@ -1674,6 +1674,13 @@ void DocumentBroker::alertAllUsers(const std::string& msg)
 
+void DocumentBroker::setKitLogLevel(const std::string& level)
+{
+    assertCorrectThread();
+    std::unique_lock<std::mutex> lock(_mutex);
```
If it's called form the "correct thread" then we don't need the lock. See other cases where we access _childProcess and also the (few) cases where the lock is used.

gokaysatir replied:
I removed lock.

Ashod reviewed (commented):
Thanks @gokaysatir! Looks much better. Left some notes.

gokaysatir commented:
@Ashod thanks for the comments. I sent a new commit.

Ashod commented on wsd/Admin.cpp:
```diff
@@ -603,42 +603,32 @@ unsigned Admin::getNetStatsInterval()
 }
 
-void Admin::setChannelLogLevel(const std::string& _channelName, std::string _level)
+void Admin::setChannelLogLevel(const std::string& _channelName, std::string level)
```
Good, `_level` should be `level`, perhaps do the same with `_channelName` ?

gokaysatir replied:
Thanks for the notes. I replaced _channelName variable.

Ashod reviewed (commented):
Thanks @gokaysatir! 

I don't see any issues, but I wonder why this isn't done on Master first, instead of a stable branch.

Normally, we get new features added to Master, then we port to the release branch(es), if necessary.

I'm very hesitant to merge a new feature into a release branch before it gets vetted in Master.

A couple of minor notes: reviewers don't mark every case/line for an issue. Authors should check the patch for similar cases as the one mentioned in a review issue. You fixed many of the issues I raised, thank you, but you left other similar cases. Also, best to run `make check` (the checklist in the description suggests that you haven't).

Otherwise, thanks for the patch, it looks good.
